crates/libtiny_tui/src/key_map.rs

```diff
@@ -0,0 +1,295 @@
+    }
+
+    pub(crate) fn load(&mut self, key_map: &KeyMap) {
+        self.0.extend(key_map.0.iter())
+    }
+}
+
+impl<'de> Deserialize<'de> for KeyMap {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct KeyMapVisitor;
+
+        impl<'de> Visitor<'de> for KeyMapVisitor {
+            type Value = KeyMap;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(
+                    formatter,
+                    "a map of key bindings. ex. 'ctrl_a: input_move_curs_start'"
+                )?;
+                write!(formatter, "defaults: \n{}", KeyMap::default())
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: MapAccess<'de>,
+            {
+                let mut key_map = HashMap::with_capacity(map.size_hint().unwrap_or(0));
+                while let Some((key, action)) = map.next_entry::<MappedKey, KeyAction>()? {
+                    key_map.insert(key.0, action);
+                }
+                Ok(KeyMap(key_map))
+            }
+        }
+        deserializer.deserialize_map(KeyMapVisitor)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MappedKey(Key);
+
+impl<'de> Deserialize<'de> for MappedKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct MappedKeyVisitor;
+
+        impl<'de> Visitor<'de> for MappedKeyVisitor {
+            type Value = MappedKey;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(formatter, "single keys: backspace, del, end, esc, home, pgdown, pgup, tab, up, down, left right, [a-z], [0-9]. ")?;
+                write!(
+                    formatter,
+                    "modifiers with arrow key or single characters:  alt, shift, ctrl"
+                )
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let key_combo = v.split_once('_');
+                let key = match key_combo {
+                    None => match v {
+                        "backspace" => Key::Backspace,
+                        "del" => Key::Del,
+                        "end" => Key::End,
+                        "esc" => Key::Esc,
+                        "home" => Key::Home,
+                        "pgdown" => Key::PageDown,
+                        "pgup" => Key::PageUp,
+                        "tab" => Key::Tab,
+                        "up" => Key::Arrow(Arrow::Up),
+                        "down" => Key::Arrow(Arrow::Down),
+                        "left" => Key::Arrow(Arrow::Left),
+                        "right" => Key::Arrow(Arrow::Right),
+                        ch if ch.chars().size_hint().0 == 1 => {
```

Comment:
I think `size_hint` is dangerous in cases like this, quoting the documentation:

> size_hint() is primarily intended to be used for optimizations such as reserving space for the elements of the iterator
> ...
> The default implementation returns (0, None) which is correct for any iterator.

So in principle this could return `0` even though there is one character in the iterator.

For `Chars` generated from a `str` it probably returns the right value, but this kind of generic code often does not survive refactoring and cause bugs in the long term. In this case we could imagine using another iterator instead of `str::chars`. `size_hint` would still work (as in it type checks), but depending on the implementation this code may break.

We should also replace other uses of `size_hint` in the PR, for the same reason.